### PR TITLE
rule to add username to metadata Forrestor

### PR DESCRIPTION
**Pull request by marygreenleeAUTH0**
By submitting a PR to this repository, you agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md). Please see the [contributing guidelines](https://github.com/auth0/.github/blob/master/CONTRIBUTING.md) for how to create and submit a high-quality PR for this repo.

### Description
This rule is for the Forrester CIAM project. This rule adds username to the app_metadata of a user on first login. The purpose of this is so the username can be put into an email template using Liquid syntax

### References

> Include any links supporting this change such as a:
>
> - GitHub Issue/PR number addressed or fixed
> - Auth0 Community post
> - StackOverflow post
> - Support forum thread
> - Related pull requests/issues from other repos
>
> If there are no references, simply delete this section.

### Testing

> Describe how this can be tested by reviewers. Be specific about anything not tested and reasons why. If this library has unit and/or integration testing, tests should be added for new functionality and existing tests should complete without errors.
>
> Please include any manual steps for testing end-to-end or functionality not covered by unit/integration tests.
>
> Also include details of the environment this PR was developed in (language/platform/browser version).

- [ ] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [ ] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [ ] All active GitHub checks for tests, formatting, and security are passing
- [ ] The correct base branch is being used, if not `master`

### CODE REVIEW

File: src/rules/Username-attribute.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * This is useful for cases where you want to add the username to an email using liquid syntax.
+ */
+ 
+ function addPersistenceAttribute(user, context, callback) {
+    user.app_metadata = user.app_metadata || {};
+    // short-circuit if the user signed up already
+    if (user.app_metadata.username) return callback(null, user, context);
+    // first time login/signup
+    user.app_metadata.username = user.app_metadata.username || user.username;
+    context.idToken['https://example.com/favorite_color'] = user.app_metadata.username;
```
Comment (seejee) [inline src/rules/Username-attribute.js:18]: This is setting a `favorite_color` property. Should it be something like `/username` instead?

File: src/rules/Username-attribute.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    auth0.users.updateAppMetadata(user.user_id, user.app_metadata)
+      .then(function(){
+          callback(null, user, context);
+      })
+      .catch(function(err){
+          callback(err);
+      });
+  }
```
Comment (seejee) [inline src/rules/Username-attribute.js:26]: Typically the rules are called something like "set-username-attribute" (all lowercase). Mind rename this file to match that format?

File: test/rules/Username-attribute.test.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
+'use strict';
+
+const loadRule = require('../utils/load-rule');
+const ContextBuilder = require('../utils/contextBuilder');
+const RequestBuilder = require('../utils/requestBuilder');
+
+const ruleName = 'add-country';
```
Comment (seejee) [inline test/rules/Username-attribute.test.js:7]: This is executing the wrong rule. To execute your rule, please change this to the rule filename that you settle on for the file above.

File: src/rules/Username-attribute.js (hunk 1 above)
Comment (marygreenleeAUTH0) [reply src/rules/Username-attribute.js:18]: I will get this fixed

File: src/rules/Username-attribute.js (hunk 2 above)
Comment (marygreenleeAUTH0) [reply src/rules/Username-attribute.js:26]: I will fix and repushh

File: src/rules/Username-attribute.js (hunk 1 above)
Comment (marygreenleeAUTH0) [reply src/rules/Username-attribute.js]: @seejee I fixed this

File: src/rules/Username-attribute.js (hunk 2 above)
Comment (marygreenleeAUTH0) [reply src/rules/Username-attribute.js]: @seejee Fixed

File: test/rules/Username-attribute.test.js (hunk 3 above)
Comment (marygreenleeAUTH0) [reply test/rules/Username-attribute.test.js]: Fixed

**Pull request merged**